File: 백준/Gold/12919. A와 B 2/A와 B 2.cc
#include <bits/stdc++.h>
using namespace std;
string s, t;
bool flag = false;

void dfs(string cur) {
    if(cur == s) { 
        flag = true;
        return;
    }
    if(cur.size() < s.size()) return;
    if(!cur.empty() && cur.back() == 'A') {
        string next = cur;
        next.pop_back();
        dfs(next);
    }
    if(!cur.empty() && cur.front() == 'B') {
        string next = cur;
        reverse(next.begin(), next.end());
        next.pop_back();
        dfs(next);
    }
}

int main() {
    cin >> s;
    cin >> t;
    dfs(t);
    if(flag) cout << 1;
    else cout << 0;
}

File: 백준/Gold/12919. A와 B 2/A와 B 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
    static String s, t;
    static boolean flag = false;
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        s = br.readLine();
        t = br.readLine();

        dfs(t);
        if(flag) System.out.println(1);
        else System.out.println(0);
    }

    static void dfs(String cur) {
        if(cur.equals(s)) {
            flag = true;
            return;
        }

        if(cur.length() < s.length()) {
            return;
        }

        // 마지막이 a면 삭제
        if(cur.charAt(cur.length()-1) == 'A') {
            String next = cur.substring(0, cur.length()-1);
            dfs(next);
        }

        // 마지막이 b면 뒤집고 삭제
        if(cur.charAt(0) == 'B') {
            StringBuilder sb = new StringBuilder(cur).reverse();
            String next = sb.toString().substring(0, cur.length()-1);
            dfs(next);
        }
    }
}